hammers/scripts/maintenance_reservation.py

```diff
@@ -1,144 +1,224 @@
+        resource_properties = '["=", "$uid", "{node_uuid}"]'.format(node_uuid=node.uuid)
+        phys_res = {'min': "1", 'max': "1", 'hypervisor_properties': "", 'resource_properties': resource_properties, 'resource_type': 'physical:host'}
+        lease = blazar.lease.create(name=MAINT_LEASE_NAME.format(node_name=node.name.replace(' ', '_'), operator=operator.replace(' ', '_'), reason=reason.replace(' ', '_')),
+                                    start=start_time.strftime('%Y-%m-%d %H:%M'),
+                                    end=end_time.strftime('%Y-%m-%d %H:%M'),
+                                    reservations=[phys_res],
+                                    events=[])
+        print("Lease {name} (id: {id}) created successfully!".format(name=lease['name'], id=lease['id']))
+    
+    return start_time_str_in_ct, end_time_str_in_ct
 
+@prometheus_exporter(__file__)
+def main(argv=None):
+    if argv is None:
+        argv = sys.argv
+
+    parser = argparse.ArgumentParser(description='Reserve nodes for maintenance')
+    append_global_identity_args(parser, argv)
+    
+    mysqlargs = MySqlArgs({
+        'user': 'root',
+        'password': '',
+        'host': 'localhost',
+        'port': 3306,
+    })
+    mysqlargs.inject(parser)
+    
+    parser.add_argument('--operator', type=str, required=True, help='Chameleon account username of the operator')
+    parser.add_argument('--nodes', type=str, required=True, help='node ids or node names; comma separated')
+    parser.add_argument('--reason', type=str, required=True, help='maintenance reasons')
+    parser.add_argument('--dry-run', action="store_true", help='perform a trial run without making reservations')
+    parser.add_argument('--start-time', type=valid_date, default=None, help='lease start time in CT (YYYY-mm-DD HH:MM:SS); if not given, start at the earliest possible datetime')
+    parser.add_argument('--estimate-hours', type=int, default=24, help='estimated hours required for maintenance; default is 24 hours')
+    parser.add_argument('--slack', type=str, default=None, help='JSON file with Slack webhook information to send a notification to')
+
+    args = parser.parse_args(argv[1:])
+    
+    # connect to database
+    mysqlargs.extract(args)
+    db = mysqlargs.connect()
+    
+    # keystone authentication
+    auth_args = {'auth_url': args.os_auth_url,
+                 'username': args.os_username,
+                 'password': args.os_password,
+                 'project_id': args.os_project_id,
+                 'project_name': args.os_project_name}
+    if args.os_user_domain_name:
+        auth_args['user_domain_name'] = args.os_user_domain_name
+    if args.os_project_domain_name:
+        auth_args['project_domain_name'] = args.os_project_domain_name
+    # get admin session for node information
+    admin_sess = get_session(**auth_args)
+    # get maint session for creating lease
+    auth_args['project_name'] = 'maintenance'
```

Comment:
I don't think you need name if you are already using id.

Reply:
Also, why not use `args.os_project_id`? Because you always want to ensure it gets done via a maintenance user? I'm a bit surprised just changing the project here works, but I guess if you're an admin...

Reply:
> Also, why not use `args.os_project_id`? Because you always want to ensure it gets done via a maintenance user? I'm a bit surprised just changing the project here works, but I guess if you're an admin...

I think you need to be admin to get node information. In the google doc, I put `source adminrc` before running the command.